File: src/test/scala/parse/dont/validate/NIRMutatedProperties.scala
package parse.dont.validate

import org.scalacheck.Gen.*
import org.scalacheck.Prop.{classify, collect, falsified, forAll}
import org.scalacheck.{Arbitrary, Gen, Properties}
import org.scalatest.OptionValues
import parse.dont.validate.Generators.validNIR
import parse.dont.validate.NIRDomain.*
import parse.dont.validate.NIRDomain.NIR.parseNIR

object NIRMutatedProperties
    extends Properties("Mutated NIR")
    with OptionValues {

  private case class Mutator(name: String, func: NIR => Gen[String]) {
    def mutate(nir: NIR): String = func(nir).sample.value
  }

  private def mutantGenerator: Gen[Mutator] =
    oneOf(
      sexMutator,
      yearMutator,
      monthMutator,
      departmentMutator,
      cityMutator,
      serialNumberMutator,
      keyMutator,
      truncateMutator
    )

  private val sexMutator: Mutator =
    Mutator(
      "sex mutator",
      nir =>
        choose(3, 9)
          .map(s => s + nir.toString.substring(1))
    )

  private val yearMutator: Mutator =
    Mutator(
      "year mutator",
      nir =>
        frequency(
          (7, choose(100, 999)),
          (3, choose(1, 9))
        ).map(y => nir.toString.take(1) + y + nir.toString.substring(3))
    )

  private val monthMutator: Mutator =
    Mutator(
      "month mutator",
      nir =>
        choose(13, 99)
          .map(m => nir.toString.take(3) + m + nir.toString.substring(5))
    )

  private val departmentMutator: Mutator =
    Mutator(
      "department mutator",
      nir =>
        frequency(
          (9, choose(100, 999)),
          (1, choose(96, 98))
        ).map(d => nir.toString.take(5) + d + nir.toString.substring(7))
    )

  private val cityMutator: Mutator =
    Mutator(
      "city mutator",
      nir =>
        digits3Gen.map(c =>
          nir.toString.take(7) + c + nir.toString.substring(10)
        )
    )

  private val serialNumberMutator: Mutator =
    Mutator(
      "serial mutator",
      nir =>
        digits3Gen.map(s =>
          nir.toString.take(10) + s + nir.toString.substring(13)
        )
    )

  private val keyMutator: Mutator =
    Mutator(
      "key mutator",
      nir =>
        oneOf(
          (0 to 97)
            .filter(x => x != nir.calculateKey().value)
        ).map(k => nir.toString.take(13) + k)
    )

  private val truncateMutator: Mutator =
    Mutator(
      "truncate mutator",
      nir =>
        choose(1, 14)
          .map(size => nir.toString.take(size))
    )

  private val digits3Gen =
    frequency(
      (7, choose(1000, 9999)),
      (3, choose(1, 99))
    )

  property("can never be parsed") = forAll(validNIR, mutantGenerator) {
    (nir, mutator) =>
      {
        classify(true, mutator.name) {
          parseNIR(mutator.mutate(nir)).isLeft
        }
      }
  }
}

File: solution/f#/Nir-Kata-Tests/Parse.Dont.Validate/NIRMutatedProperties.fs
module NIRMutatedProperties

open FsCheck
open FsCheck.Xunit
open Microsoft.FSharp.Core
open NirGenerator
open Nir_Kata.Parse.Dont.Validate.NIRDomain.Key
open Nir_Kata.Parse.Dont.Validate.NIRDomain.NIRDomain

type Mutator =
    { name: string
      func: NIR -> Gen<string> }

let private digits3Gen =
    Gen.frequency [ (7, Gen.choose (1000, 9999))
                    (3, Gen.choose (1, 99)) ]

let private sexMutator: Mutator =
    { name = "Sex mutator"
      func =
        fun nir ->
            Gen.choose (3, 9)
            |> Gen.map (fun s -> s |> string)
            |> Gen.map (fun invalidSex -> invalidSex + nir.ToString().[1..]) }

let private yearMutator: Mutator =
    { name = "Year mutator"
      func =
        fun nir ->
            Gen.frequency [ (7, Gen.choose (100, 999))
                            (3, Gen.choose (1, 9)) ]
            |> Gen.map (fun s -> s |> string)
            |> Gen.map (fun invalidYear -> $"{nir.ToString().[0]}{invalidYear}{nir.ToString().[3..]}") }

let private monthMutator: Mutator =
    { name = "Month mutator"
      func =
        fun nir ->
            Gen.choose (13, 99)
            |> Gen.map (fun s -> s |> string)
            |> Gen.map (fun invalidMonth -> $"{nir.ToString().[0..2]}{invalidMonth}{nir.ToString().[5..]}") }

let private departmentMutator: Mutator =
    { name = "Department mutator"
      func =
        fun nir ->
            Gen.frequency [ (7, Gen.choose (100, 999))
                            (3, Gen.choose (96, 98)) ]
            |> Gen.map (fun s -> s |> string)
            |> Gen.map (fun invalidDepartment -> $"{nir.ToString().[0..4]}{invalidDepartment}{nir.ToString().[7..]}") }

let private cityMutator: Mutator =
    { name = "City mutator"
      func =
        fun nir ->
            digits3Gen
            |> Gen.map (fun s -> s |> string)
            |> Gen.map (fun invalidCity -> $"{nir.ToString().[0..6]}{invalidCity}{nir.ToString().[10..]}") }

let private serialNumberMutator: Mutator =
    { name = "Serial Number mutator"
      func =
        fun nir ->
            digits3Gen
            |> Gen.map (fun s -> s |> string)
            |> Gen.map (fun invalidSerialNumber ->
                $"{nir.ToString().[0..9]}{invalidSerialNumber}{nir.ToString().[13..]}") }

let private keyMutator: Mutator =
    { name = "Key mutator"
      func =
        fun nir ->
            Gen.choose (0, 97)
            |> Gen.filter (fun x -> Key x <> nir.calculateKey.Value)
            |> Gen.map (fun s -> s |> string)
            |> Gen.map (fun invalidKey -> $"{nir.ToString().[0..12]}{invalidKey}") }

let private truncateMutator: Mutator =
    { name = "Truncate mutator"
      func =
        fun nir ->
            Gen.choose (1, 13)
            |> Gen.map (fun size -> nir.ToString().[0..size]) }

type MutatorGenerator =
    static member mutate() : Arbitrary<Mutator> =
        Gen.elements [ sexMutator
                       yearMutator
                       monthMutator
                       departmentMutator
                       cityMutator
                       serialNumberMutator
                       keyMutator
                       truncateMutator ]
        |> Arb.fromGen

let private shouldBeError =
    function
    | Error _ -> ()
    | _ -> failwith "Should not be OK"

[<Property(Arbitrary = [| typeof<NirGenerator>
                          typeof<MutatorGenerator> |])>]
let ``invalid nir can never be parsed`` (nir: NIR, mutator: Mutator) =
    mutator.func nir
    |> Gen.sample 0 1
    |> List.head
    |> parseNIR
    |> shouldBeError
    |> Prop.classify true mutator.name
